eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs

```diff
@@ -752,4 +776,156 @@ electBasicCommittee = do
 logCurPParams :: (EraGov era, ToExpr (PParamsHKD Identity era)) => ImpTestM era ()
 logCurPParams = do
   pp <- getsNES $ nesEsL . curPParamsEpochStateL
-  logEntry $ "Current PParams:\n--------------" <> showExpr pp <> "\n--------------"
+  logEntry $
+    unlines
+      [ ""
+      , "----- Current PParams -----"
+      , showExpr pp
+      , "---------------------------"
+      , ""
+      ]
+
+submitInitConstitutionGovAction ::
+  (ShelleyEraImp era, ConwayEraTxBody era) =>
+  ImpTestM era (GovActionId (EraCrypto era))
+submitInitConstitutionGovAction = do
+  submitConstitutionGovAction SNothing
+
+submitChildConstitutionGovAction ::
+  (ShelleyEraImp era, ConwayEraTxBody era) =>
+  GovActionId (EraCrypto era) ->
+  ImpTestM era (GovActionId (EraCrypto era))
+submitChildConstitutionGovAction gai =
+  submitConstitutionGovAction $ SJust $ GovPurposeId gai
+
+submitConstitutionGovAction ::
+  (ShelleyEraImp era, ConwayEraTxBody era) =>
+  StrictMaybe (GovPurposeId 'ConstitutionPurpose era) ->
+  ImpTestM era (GovActionId (EraCrypto era))
+submitConstitutionGovAction pgai = do
+  constitutionHash <- freshSafeHash
+  let constitutionAction =
+        NewConstitution
+          pgai
+          ( Constitution
+              ( Anchor
+                  (fromJust $ textToUrl 64 "constitution.dummy.0")
+                  constitutionHash
+              )
+              SNothing
+          )
+  submitGovAction constitutionAction
+
+data PTrees a = PTrees
+  { ptPParamUpdate :: Tree a
+  , ptHardFork :: Tree a
+  , ptCommittee :: Tree a
+  , ptConstitution :: Tree a
+  }
+  deriving (Show, Eq, Generic, Default)
+
+pattern E :: Forest () -> Tree ()
+pattern E a <- Node () a
+  where
+    E a = Node () a
+
+pattern N ::
+  forall era.
+  era ~ Conway =>
+  GovActionId (EraCrypto era) ->
+  Forest (GovActionId (EraCrypto era)) ->
+  Tree (GovActionId (EraCrypto era))
+pattern N a b <- Node a b
+  where
+    N a b = Node a b
+
+pattern I ::
+  forall era.
+  era ~ Conway =>
+  Forest (StrictMaybe (GovActionId (EraCrypto era))) ->
+  Tree (StrictMaybe (GovActionId (EraCrypto era)))
+pattern I a <- Node SNothing a
+  where
+    I a = Node SNothing a
+
+pattern J ::
+  forall era.
+  era ~ Conway =>
```

Comment:
Why is this constraint necessary?

Reply:
I have changed this to directly use `Conway` instead of a type-variable. For  some reason, and I tried fixing it with the double constraints for pattern synonyms as well, these result in GHC attempting to unify two different variables.

Reply:
I got rid of the pattern synonyms entirely. :+1: